[TDD - 자동차 경주] 1단계 미션 제출합니다.

bigpel66 opened the pull request:
페어로 작성한 미션을 개별로 제출해야함을 놓쳐 이제 제출하게 되었습니다.

owenoweny committed:
feat: 움직임 기능과 테스트
- Car: 자동차 위치 및 이름, 비교 가능 여부
- MoveChecker: 자동차 움직임 가능 여부 판단
- RandomGenerator: 0-9 사이 난수 생성

owenoweny committed:
feat: 게임 진행 기능 구현

- Game.java: 게임 진행 로직 구현
- Cars.java: Car 일급 컬렉션 클래스 구현
- 그 외 검증 로직 구현

owenoweny committed:
feat: 게임 진행 상태 및 결과 출력

- Cars.java: 우승자 판별
- Game.java -> RacingCarGame.java
- RacingCarGame.java: 시도 회차별 상태 구문
- RacingCarApplication.java: 게임 진행 로직 구현

owenoweny committed:
refactor: 게임 진행 로직과 뷰를 분리

- ConsoleView.java: 각종 인풋과 결과 프롬프트 출력
- RacingCarApplication.java: 코드 스타일을 컨벤션에 맞게 수정

owenoweny committed:
refactor: 패키지 구조 수정

owenoweny committed:
feat: 문자열 입력 파싱 기능 구현

owenoweny committed:
refactor: 파서 구조 리팩토링

- Parser.java: 파서가 상태를 가지게 변경

owenoweny committed:
feat: 덧셈 기능 구현

owenoweny committed:
feat: 계산기 실행 기능 구현

- Parser 생성자 접근 제어자 수정

owenoweny committed:
feat: 계산기 파서에서 정규표현식 예약어를 필터링

owenoweny committed:
feat: 계산기 파서에서 음수 입력을 필터링

owenoweny committed:
docs: 기능 목록 수정

owenoweny committed:
refactor: 패키지 구조 수정, 코드 스타일 통일

owenoweny committed:
fix: CalculatorApp - 입력 문자열을 입력받아 합 정수를 반환하는 형태로 메소드 수정

Commit:
fix: 1차 피드백을 위한 수정

Commit:
chore: 임시로 두었던 테스트 삭제

Commit:
fix: iteration 변수 이름을 단순화

Commit:
chore: delete calculator package

Commit:
feat: Adder의 sum 메서드 추가

Commit:
refactor: Calculator 실행 로직 및 검증 개선
- Calculator 앱 이름 변경
- main 실행에서 메서드 실행으로 변경함에 따라 인자 검증에 사용되는 메서드 원형을 변경
- 요구사항에 나타난대로 RuntimeException이 발생 시 0을 반환

Commit:
feat: Numbers 객체 작성

Commit:
test: Calculator 객체 테스트
- 0이 반환 되는 (null or 빈 문자열) 경우를 확인

Commit:
feat: Parser 객체 생성
- Calculator 메인 로직 완성
- Adder가 List 형태의 인풋을 이용하지 않고 Numbers를 이용하도록 변경
- Parser의 윤곽

Commit:
feat: 문자열 덧셈 계산기 구현
- Calculator에서 NPE, IllegalArgument를 잡아서 0을 반환
- Number의 Equals 및 HashCode로 동치 파악
- Parser 로직 완성 및 예외 처리
- 각 객체 테스트 코드 보완

Commit:
chore: RacingGame 삭제

Commit:
docs: 덧셈 계산기 기능 목록 작성

liquidjoo commented on src/main/java/calculator/domains/Adder.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.List;
+
+public class Adder {
+    public static int sum(List<Integer> input) {
```
정적메서드를 사용한다면 해당 클래스가 의미있는 클래스일까요?
불완전객체에 대해서 한번 알아보면 어떨까요

bigpel66 replied:
이 부분은 CalculatorApp과 Adder의 설계가 모호한 것도 한 몫한 거 같아요. 계산만 담당하는 Calculator, 문자열 입력 값을 숫자로 갖는 Wrapping 객체 Numbers, 입력 값 파싱을 수행하는 Parser와 같이 고쳐서 Adder의 기능을 조금 더 의미 있은 곳으로 옮겨보려고 합니디.

liquidjoo commented on src/main/java/calculator/domains/Numbers.java:
```diff
@@ -0,0 +1,56 @@
+        } catch (NumberFormatException e) {
+            throw new RuntimeException("문자를 입력할 수 없습니다.");
+        }
+        validateNegatives();
```
numbers 를 생성하는 로직을 합쳐 볼 수 있을 것 같은데요
로직이 길어진다면 메서드로 분리해서 만들어 보면 좋을 것 같아요

liquidjoo commented on src/main/java/calculator/domains/Numbers.java:
```diff
@@ -0,0 +1,56 @@
+    public static Numbers of(String input) {
+        validateRegex(input);
+        if (!input.startsWith("//")) {
+            return new Numbers(",|:", input);
```
문자열은 상수화 시켜 보면 코드의 가독성이 좋아질 것 같네요

liquidjoo commented on src/main/java/calculator/domains/Numbers.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.stream.Collectors;
+
+public class Numbers {
+    private static final Pattern defaultInputFormatPattern = Pattern.compile("//(.)\n(.*)");
+    private static final Pattern forbiddenDelimiterPattern = Pattern.compile("\\.|\\^|\\$|\\*|\\+|\\?|\\|");
```
객체의 메모리 효율성 👍 
> 저는 개인적으로 static final 을 통해 정의한 것들은 상수로 여기며 대문자를 주로 사용합니다 :)

liquidjoo commented on src/main/java/calculator/domains/Numbers.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public List<Integer> find() {
+        return numbers;
```
방어적 복사에 대해서 알아보고 적용해보면 좋을 것 같아요 :)

bigpel66 replied:
Numbers가 자신을 표현하는 일급 콜렉션을 감싼 형태로 구성했는데요. 방어적 복사가 불변성을 위한 개념으로 이해하고 있어요. 말씀해주신 부분을 Collections.unmodifiableList 이용하는 식으로 수정해볼게요.

liquidjoo commented on src/main/java/racinggame/ConsoleView.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public static void printCarOffset(RacingCarGame racingCarGame) {
+        racingCarGame.showStatus();
```
View 에 해당하는 영역일 것 같아요 :)
객체 값을 전달해 View 에서 표현해보면 좋을 것 같아요

liquidjoo commented on src/main/java/racinggame/RacingCarApplication.java:
```diff
@@ -0,0 +1,27 @@
+
+        ConsoleView.printResultPrompt();
+        for (int i = 0; i < trialInput; i++) {
+            racingCarGame.processTurn();
+            ConsoleView.printCarOffset(racingCarGame);
```
결과를 실행 후 화면에 노출하는 것도 좋으나
비즈니스로직을 수행 후 결과 값을 화면에 노출해보아요 :)

liquidjoo commented on src/main/java/racinggame/domains/Car.java:
```diff
@@ -0,0 +1,55 @@
+    private final String owner;
+    private int offset;
+
+    public Car(String owner, int offset) {
+        validateLength(owner);
+        this.owner = owner;
+        this.offset = offset;
+    }
+
+    public static Car of(String owner) {
+        return new Car(owner, 0);
+    }
```
정적팩토리 메서드 패턴을 사용한다면 기본 생성자는 접근제어자를 통해 제어해주는 것이 어떨까요?

liquidjoo commented on src/main/java/racinggame/domains/Car.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    public void move() {
+        this.offset++;
+    }
```
void 타입도 좋으나 명확한 타입반환을 통해 직관적으로 표현해 보는 것이 어떨까요?

bigpel66 replied:
int를 반환하는 식을 말씀하시는거라면, int가 반환 되었을 때 어떤 용례를 의식하셨는지 궁금합니다.

liquidjoo commented on src/main/java/racinggame/domains/Cars.java:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    public static Cars of(List<String> carNames) {
+        return new Cars(carNames.stream().map(Car::of).collect(Collectors.toList()));
+    }
```
Car 와 마찬지로 기본 생성자에 대한 접근제어자를 활용해 보아요 :)

liquidjoo commented on src/main/java/racinggame/domains/Cars.java:
```diff
@@ -0,0 +1,36 @@
+        return cars;
+    }
+
+    public List<Car> winners() {
+        cars.sort(Comparator.reverseOrder());
+        return cars
+                .stream()
+                .filter(e -> cars.get(0).offset() <= e.offset())
+                .collect(Collectors.toList());
+    }
```
별도의 역할로 나누어 보면 어떨까요?
어떤 장점이 있을까요?

liquidjoo commented on src/main/java/racinggame/domains/RacingCarGame.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.List;
+
+public final class RacingCarGame {
```
테스트 코드가 부족해 보여요 :)

liquidjoo commented on src/main/java/racinggame/domains/RacingCarGame.java:
```diff
@@ -0,0 +1,41 @@
+
+    public RacingCarGame(Cars cars) {
+        this.cars = cars;
+        this.randomGenerator = new RandomGenerator();
```
테스트하기 어려운 것을 어떻게 테스트 해볼 수 있을까요?

bigpel66 replied:
주입 받거나, 메서드로 분리해내서 Overriding을 이용하는 방식으로 생각하고 있어요. 반영해서 수정해두겠습니다.

liquidjoo commented on src/main/java/racinggame/domains/RacingCarGame.java:
```diff
@@ -0,0 +1,41 @@
+        this.randomGenerator = new RandomGenerator();
+    }
+
+    public RacingCarGame() {
+    }
```
다른 곳에서 사용하지 않는다면 접근제어자를 활용해 보아요 :)

liquidjoo commented on src/main/java/racinggame/domains/RacingCarGame.java:
```diff
@@ -0,0 +1,41 @@
+        return this.cars;
+    }
+
+    public void processTurn() {
+        cars.list().stream()
+                .filter(e -> MoveChecker.check(randomGenerator.generate()))
+                .forEach(Car::move);
+    }
```
테스트 코드를 작성해 보면 좋을 것 같아요 :)

liquidjoo commented on src/main/java/racinggame/utils/RandomGenerator.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Random;
+
+public class RandomGenerator {
+    private final Random random;
+
+    public RandomGenerator() {
+        random = new Random();
+    }
```
랜덤 객체의 생성비용을 줄여볼 수 있을 것 같네요 :)

liquidjoo commented on src/test/java/racinggame/domains/CarsTest.java:
```diff
@@ -0,0 +1,31 @@
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+
+@DisplayName("게임에 참가하는 자동차들 관련 테스트")
+public class CarsTest {
```
winners 에 대한 테스트도 있으면 좋을 것 같아요 :)
> 책임 이동이 발생한다면 해당 책임에 대한 테스트를 진행해주시면 됩니다

liquidjoo commented on src/test/java/racinggame/domains/CarTest.java:
```diff
@@ -0,0 +1,46 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@DisplayName("자동차 객체 테스트")
+public class CarTest {
```
compareTo 를 사용해준 만큼 해당 로직에 대한 테스트도 존재하면 좋을 것 같아요~

liquidjoo commented on src/test/java/racinggame/domains/RacingCarGameTest.java:
```diff
@@ -0,0 +1,19 @@
+
+
+@DisplayName("")
+public class RacingCarGameTest {
```
비즈니스 로직 즉 자동차의 움직임을 담당하고 있는 객체인데 주요한 테스트 코드가 없는 것 같아요~

liquidjoo reviewed (changes requested):
안녕하세요 종환님
이번 미션 리뷰어를 맡은 김성주 입니다 :)
미션 잘 구현해주셨어요 👍
몇 가지 코멘트 남겼는데요~ 확인하고 다시 요청주세요 🙇

bigpel66 commented:
@liquidjoo 페어와 별개로 제출해야하는 부분을 놓치면서 일정이 밀리다보니, 눈에 띄는 부분이 있었지만 리뷰 요청을 급하게 먼저 드리게 되었네요. 그러다보니 수정할 부분들이 많이 드러나기도 했고, 대체로 공감하는 부분들입니다. 늦은 시간에 리뷰 감사드리고, 언급해주신 부분들은 적절히 반영해서 다시 요청 드려보겠습니다.

감사합니다.

Commit:
fix: 1차 피드백 반영
- Adder 로직을 Calculator와 병합
- Numbers 내부 요소의 불변성을 보장
- Numbers 동치 테스트 추가
- Parser 내부 요소 이름 컨벤션 준수

Commit:
feat: Car 객체 생성
- 자동차가 이름을 갖도록 설정 및 이름의 입력 값의 길이 등을 검증
- 자동차가 4이상의 값에서 전진 가능
- 자동차의 상태는 이름과 위치 값으로 표현하도록 구성

Commit:
fix: 함수의 수직 배치를 호출 순서로 변경

Commit:
feat: Cars 객체 작성
- Car의 동치를 위해 equals와 hashCode 추가
- Cars에 등록된 자동차들의 중복을 검사
- Cars를 이용하여 일괄 전진 기능 추가
- Cars의 중복 참여, 생성 여부, 전진 기능을 테스트

Commit:
feat: 레이싱 게임 구성 완료
- Application.java: ConsoleView 입력 값을 이용하여 RacingGame을 진행하도록 작성
- Car.java: 자동차의 이름과 예외 메시지 변경
- Cars.java: 관리 대상인 자동차들의 상태를 일괄적으로 얻도록 statusAll 작성
- ConsoleView.java: 게임 진행에 필요한 입력 값을 구성
- RacingGame.java: 입력 값을 이용하여 생성된 Cars와 trial을 이용하여 게임을 진행
- 각 테스트와 기능 목록을 작성

Commit:
fix: checkNegative -> checkNonNegative

liquidjoo commented on README.md:
```diff
@@ -11,28 +11,31 @@
+- [x] Parser는 입력 값을 적절히 파싱
+    - [x] Default인지 Custom인지 판별
+    - [x] 파싱 후 Numbers를 반환
+    - [x] 입력 값의 Blank 여부를 파악하여 예외 처리
+    - [x] 파싱 후 정수 값으로의 변환에서의 예외 처리
+- [x] Calculator는 sum 로직을 제공
+    - [x] 입력 값의 Blank 여부를 파악하여 예외 처리
+    - [x] 입력 값이 정수가 아닌 경우와 음수인 경우 예외 처리
 
-# 문자열 덧셈 계산기
+# 레이싱 게임
 
 ## 기능 목록
 
-- 수식을 입력받는다.
-    - 음수 또는 숫자가 아닌 값인 경우, RumtimeException을 발생시킨다.
-- 계산 결과를 출력한다.
-    - null 또는 빈 문자열인 경우 0을 출력한다.
-- 수식을 파싱한다.
-    - ',', ':'를 구분자로 정수 문자열의 합을 계산한다.
-    - 커스텀 구분자가 있는 경우, 이를 구분자로 갖는 정수 문자열의 합을 계산한다.
-    - 정규표현식의 예악어는 커스텀 구분자가 될 수 없다.
+- [x] 자동차의 이름은 5글자 이하
+- [x] 자동차는 전진 가능
+- [x] 자동차의 상태를 파악
+- [x] 자동차들을 묶은 객체에서 중복 참여를 검증
+- [x] 자동차들을 묶은 객체를 이용한 전진
+- [x] 자동차들을 묶은 객체를 이용한 상태 파악
+- [x] 자동차들과 시도 횟수를 이용하여 게임을 진행하는 게임 객체 작성
+- [x] 메시지를 출력하면서 사용자로부터 입력 값을 구성할 수 있도록 객체 작성
+- [x] 입력 값 구성 객체와 게임 객체를 이용하는 어플리케이션 작성
```
명확한 요구 사항 정리 👍

liquidjoo commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Objects;
+
+public final class Calculator {
+    private Calculator() {
+        throw new AssertionError("Calculator는 상태를 갖지 않는 유틸 객체입니다.");
+    }
```
유틸성 객체 생성 시 이러한 방법도 좋은 방법 중 하나인 것 같네요 👍

liquidjoo commented on src/main/java/calculator/Numbers.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public List<Integer> list() {
+        return list;
```
방어적 복사에 대해서 알아보면 좋습니다

liquidjoo commented on src/main/java/calculator/Parser.java:
```diff
@@ -0,0 +1,76 @@
+    private static final String CUSTOM_DELIMITER_PREFIX = "//";
+    private static final String DEFAULT_DELIMITER = ",|:";
+    private static final int DELIMITER_CAPTURE_POSITION = 1;
+    private static final int TARGET_CAPTURE_POSITION = 2;
+    private final String input;
```
상수와 필드는 나누어 주는 것이 좋습니다 :)
```suggestion
    private static final int TARGET_CAPTURE_POSITION = 2;
    
    private final String input;
```

liquidjoo commented on src/main/java/calculator/Parser.java:
```diff
@@ -0,0 +1,76 @@
+        return m;
+    }
+
+    public Numbers parse() {
+        try {
+            List<Integer> list = Arrays.stream(target.split(delimiter))
+                    .map(Integer::parseInt)
+                    .collect(Collectors.toList());
+            return Numbers.of(list);
+        } catch (NumberFormatException ex) {
+            throw new InputMismatchException("파싱 대상 문자열은 숫자가 아닙니다.");
+        }
+    }
```
Numbers 객체와 의존 관계가 생길 수 밖에 없을 것 같은데요 :)
List 를 반환 후 사용하는 곳에서 Numbers 를 생성해 보는 것은 어떨까요?

liquidjoo commented on src/main/java/racinggame/Car.java:
```diff
@@ -0,0 +1,75 @@
+        return offset;
+    }
+
+    public void move() {
+        if (pickEngineValue() >= FORWARDABLE) {
+            offset++;
+        }
+    }
```
move 에 대한 테스트를 오버라이드도 좋은 방법이지만 실질적으로 비즈니스 사용시엔 오버라이드해서 사용하는 경우가 드문 것 같아요~
인터페이스를 활용해서 move 에 대한 테스트를 진행해보면 어떨까요?

liquidjoo commented on src/main/java/racinggame/Car.java:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    protected int pickEngineValue() {
+        return ENGINE.nextInt(ENGINE_LIMIT);
+    }
```
테스트하기 어려운 테스트는 어떻게 해보면 좋을까요?

liquidjoo commented on src/main/java/racinggame/Cars.java:
```diff
@@ -0,0 +1,42 @@
+        list.forEach(Car::move);
+    }
+
+    public String statusAll() {
+        StringBuilder sb = new StringBuilder();
+        for (Car car : list) {
+            sb.append(car.status()).append('\n');
+        }
+        return sb.toString();
+    }
```
비즈니스 로직보단 View 로직에 가까워 보여요~
결과를 담고 있는 Car 객체들을 메서드 인자로 넘겨서 View 에서 그려보는 방식으로 변경해 보아요 :)

liquidjoo commented on src/main/java/racinggame/RacingGame.java:
```diff
@@ -0,0 +1,49 @@
+        return Collections.unmodifiableList(rounds);
+    }
+
+    public List<String> winners() {
+        return Collections.unmodifiableList(winners);
+    }
```
별도의 역할이 있어도 좋을 것 같아요!
이렇게 역할과 책임이 나누어진다면 어떤 장점이 있을까요?

liquidjoo reviewed (approved):
안녕하세요 종환님 :)
1단계 피드백 반영 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다~ 확인하고 다음 단계에 적용해 보면 좋을 것 같아요
다음 단계 진행해 주세요 🙇
